Support new wildcard file matching behavior

scottopell opened the pull request:
### What does this PR do?
In the current logs agent, if you configure file log sources with a wildcard path (eg, `my-log-dir/*.log`), the logs agent will resolve the wildcard every `scan` interval and decide if more files should be tailed.

There is a limit to the number of open files that is configurable, and once there are more files matching the wildcard than that limit, we need some way of deciding _which_ wildcard matches should be selected.

The goal of this PR is to provide a new matching behavior that selects based on the time a file was most recently modified.

This change provides a new config option `logs_config.file_wildcard_selection_mode` that has 2 possible values:
- `by_name` which is the existing behavior (described in more detail below)
- `by_modification_time` which is the new behavior

### Motivation
In the current behavior, the ordering is roughly alphabetical and therefore there is no way to select different files once the initial pass has been made without renaming files.

To make things worse, the existing behavior is to "greedily" fill the open file slots from each log source in turn.
This means if you have multiple wildcard log sources, say you're monitoring both `/var/serviceA/logs/*.log` and `/var/serviceB/logs/*.log`, if `/var/serviceA/logs/*.log` contains > `logs_config.open_file_limit` number of logs, `/var/serviceB/logs/*.log` will not be considered at all.

### Additional Notes
This PR creates 2 new "modes" that a `file_provider` can operate in, both allowing for a `global` mode that will consider all wildcard matches and a `mtime` mode that will order by the most recently modified file time instead of alpha ordering.

The config option `logs_config.file_wildcard_selection_mode` sets both of these options for the `file_provider`

(I changed too much of `file_provider` for git to recognize it as a rename, even when I put the rename in its own commit, sorry for the ugly diff!)

### Possible Drawbacks / Trade-offs
This mode is significantly less performant than the default. Every scan (default every 10 seconds) will now `stat` every single file that matches the glob, whereas before only the directory contents needed to be read.

Performance will be even worse in the case where the agent is monitoring a network mount of some kind, presumably those `stat` calls will be more expensive.

This is a trade-off that the user should be aware of when enabling this option.

**mbp m1 max - macos native**
```
goos: darwin
goarch: arm64
pkg: github.com/DataDog/datadog-agent/pkg/logs/internal/launchers/file/provider
BenchmarkApplyOrdering/Mtime-10                   194088              5984 ns/op
BenchmarkApplyOrdering/ReverseLexicographical-10                 7772708               158.7 ns/op
```

**mbp m1 max - linux vm**
```
goos: linux
goarch: arm64
pkg: github.com/DataDog/datadog-agent/pkg/logs/internal/launchers/file/provider
BenchmarkApplyOrdering/Mtime-4            200792              5733 ns/op
BenchmarkApplyOrdering/ReverseLexicographical-4                  7339722               166.6 ns/op
```


### Describe how to test/QA your changes
There are two goals for QA here:
1. Ensure that without the new config option set, things work as they did before
2. When there are more than `open_files_limit` matches, more recently modified files are prioritized.

Example `datadog.yaml`:
```
confd_path: /home/ubuntu/dev/datadog-config-scenarios/disk-log-listener/conf.d/
logs_enabled: true
logs_config:
  container_collect_all: false
  open_files_limit: 3
  file_wildcard_selection_mode: "by_modification_time"
  file_scan_period: 1.0
```

Example `conf.d/disk-log-listener/conf.d/conf.yaml`:
```
logs:
  - type: file
    path: /tmp/logfolder/*.log
    service: "my-service"
    source: "my-client-app"
```

1. With the config option _on_ (ie, set to `by_modification_time`), create 3 log files with some content. Ensure via `agent status` that these logs are being tailed. Then create a 4th log file with some new content. Ensure that the oldest of the 3 has been dropped and replaced by this new 4th log file.
2. With the config option _off_ (ie, set to `by_name`), create 3 log files with some content. Ensure via `agent status` that these logs are being tailed. Then create a 4th log file with a name alphabetically _less_ than the existing files. Ensure that there is no change to the set of files being tailed.


### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

scottopell committed:
Adds concept of selection strategy and sort mode to fileProvider (and tests)

scottopell committed:
Adds benchmark covering different sort modes

scottopell committed:
Moves file_provider to its own package and establishes public api

scottopell committed:
Introduces working global mode

scottopell committed:
Some refactoring and more tests around file matching

scottopell committed:
Fixes bug in launcher when openFilesLimit is met to avoid multiple 'scan's being necessary

scottopell committed:
Adds config option to enable new wildcard behavior and a few more tests

vickenty commented on pkg/config/config.go:
```diff
@@ -893,6 +893,13 @@ func InitConfig(config Config) {
+	// that match wildcard log configurations than the `logs_config.open_files_limit`
+	// Choices are 'prioritizeNewest', 'simpleAlpha'
+	// WARNING: 'prioritizeNewest' is less performant than 'simpleAlpha' and will trigger
+	// more disk I/O at the wildcard log paths
```
Would it make sense to document what these options do in more detail?

Nit pick, but I'm not a fan of these names. What if we borrow from a file manager lingo and call them `sortByName`, `sortByModifiedTime`?

scottopell replied:
Yes these should definitely be documented a bit more, is this the best place to do that?

I'm open to those new names, I was kind of avoiding `sortBy*` names because this changes the behavior slightly more than just sorting (greedy vs global log source consumption).

I guess the trade-off is a simpler name that isn't 100% accurate vs weird names that require you to check the docs.
I have no _strong_ preference here, curious what your thoughts are.

vickenty replied:
We can document them right here in the template.

I feel that global vs greedy is a relatively minor detail that doesn't necessarily need to have impact on the option name, provided that it is documented. Maybe it would sound better if we drop "sort" from the name, and call them `byName`, `byModificationTime`?

And perhaps in snake case too, for consistent style with few other enumerations already in the config.

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,281 @@
+const openFilesLimitWarningType = "open_files_limit_warning"
+
+// wildcardOrdering controls what ordering is applied to wildcard files
+type wildcardOrdering string
```
Typically such enumeration constants would be numeric, using iota. Is there a reason to make it a string? (Ditto for selectionStrategy)

djmitche replied:
Also, are these feature flags that we'd like to eventually do away with, or permanent parts of the config surface?

scottopell replied:
No particular reason, still learning how to write idiomatic go. I'll switch this to a enumeration constant.

scottopell replied:
@djmitche missed your question, but the config option that is exposed to users is intended to be permanent.

These bits are the flags that `file_provider` exposes to configure the "mode", and per Vickenty's advice I've consolidated them into a single flag.

vickenty commented on pkg/logs/internal/launchers/file/launcher.go:
```diff
@@ -60,10 +61,20 @@ type Launcher struct {
-func NewLauncher(tailingLimit int, tailerSleepDuration time.Duration, validatePodContainerID bool, scanPeriod time.Duration) *Launcher {
+func NewLauncher(tailingLimit int, tailerSleepDuration time.Duration, validatePodContainerID bool, scanPeriod time.Duration, wildcardMode string) *Launcher {
+	var prov *fileprovider.FileProvider
+	if wildcardMode == "prioritizeNewest" {
+		prov = fileprovider.NewFileProvider(tailingLimit, fileprovider.WildcardMtime, fileprovider.GlobalSelection)
+	} else if wildcardMode == "simpleAlpha" {
+		prov = fileprovider.NewFileProvider(tailingLimit, fileprovider.WildcardReverseLexicographical, fileprovider.GreedySelection)
+	} else {
+		log.Warnf("Unknown wildcard mode specified: %q, defaulting to 'simpleAlpha' strategy.\n", wildcardMode)
+		prov = fileprovider.NewFileProvider(tailingLimit, fileprovider.WildcardReverseLexicographical, fileprovider.GreedySelection)
+	}
```
Would it make sense to call `NewFileProvider` once, and use a this block instead to only pick appropriate wildcard mode to pass to the constructor? Perhaps `switch` would be appropriate here too.

vickenty commented on pkg/logs/internal/launchers/file/launcher.go:
```diff
@@ -60,10 +61,20 @@ type Launcher struct {
+	if wildcardMode == "prioritizeNewest" {
+		prov = fileprovider.NewFileProvider(tailingLimit, fileprovider.WildcardMtime, fileprovider.GlobalSelection)
+	} else if wildcardMode == "simpleAlpha" {
+		prov = fileprovider.NewFileProvider(tailingLimit, fileprovider.WildcardReverseLexicographical, fileprovider.GreedySelection)
```
Selection mode and wildcard mode map 1 to 1: do we really need to expose it in the constructor?

scottopell replied:
All combinations of selection mode and wildcard mode are valid and tested. Ie, you can specify `GlobalSelection` with `ReverseLexicographical` sorting.

So from "what is the public API of this component" perspective, it makes sense to me to expose both, even though we only use 2 combinations.

vickenty replied:
[Swapping](https://gist.github.com/vickenty/6b03029eab9d2ac4e5d988bc0cd3f604) `Greedy` and `Global` here does not seem to fail any tests, which means a mistake can creep in. Of course it can be covered by more tests,  but I think it is better to have minimal interfaces: fewer behaviors means less code and tests (which affects amount of future maintenance work required), and less chances to trigger the wrong behavior by mistake.

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,281 @@
+	wildcardFileCounter := newWildcardFileCounter()
+	wildcardSources := make([]*sources.LogSource, 0, len(inputSources))
+
+	consumeFiles := func(files []*tailer.File) {
```
Would it be possible to make this a regular method instead?

scottopell replied:
I could make this a regular method instead, passing the appropriate vars explicitly instead of capturing them in the closure.
Does that option seem more readable to you?

vickenty replied:
Yes, it would be more readable. It would also improve debugging: the function would have a proper name in tracebacks and profiles.

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,281 @@
+	for i := 0; i < len(inputSources); i++ {
+		source := inputSources[i]
+		isWildcardSource := config.ContainsWildcard(source.Config.Path)
+		if isWildcardSource {
+			// if global, save all wildcards for later
+			if p.selectionMode == GlobalSelection {
+				wildcardSources = append(wildcardSources, source)
+				continue
+			} // else if greedy, collect now source-by-source
```
Would it be more readable if we hoist this condition to the top (so we have one selectionMode condition, with both branches executing a for loop inside? Maybe as separate methods too?

djmitche reviewed (approved):
Looks good from a not-very-detailed skim.  Thinking forward a bit, this distinction might be made by selecting different implementations for an interface (represented as a component).

gh123man commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,281 @@
+	if p.wildcardOrder == WildcardMtime {
+		// sort paths descending by mtime
+		sort.SliceStable(files, func(i, j int) bool {
+			statI, _ := os.Stat(files[i].Path)
```
Would it be more efficient (and possibly more performant) to first call stat on all the files, and then sort them based on the results? 

This will call `os.Stat` at least twice for every file as is.

scottopell replied:
Good point! Will stat them all first and then sort over the results

gh123man commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,281 @@
+		// sort paths descending by mtime
+		sort.SliceStable(files, func(i, j int) bool {
+			statI, _ := os.Stat(files[i].Path)
+			statJ, _ := os.Stat(files[j].Path)
```
It may also be worth adding a unit test for invalid file paths/removed files.

gh123man commented on pkg/logs/internal/launchers/file/launcher.go:
```diff
@@ -60,10 +61,20 @@ type Launcher struct {
+	} else if wildcardMode == "simpleAlpha" {
+		prov = fileprovider.NewFileProvider(tailingLimit, fileprovider.WildcardReverseLexicographical, fileprovider.GreedySelection)
+	} else {
+		log.Warnf("Unknown wildcard mode specified: %q, defaulting to 'simpleAlpha' strategy.\n", wildcardMode)
```
NIT: Do we need the `\n`? Most of our other `Warnf` messages don't append a newline (but a few do).

scottopell replied:
Every log goes through the scrubber which will by default add a trailing newline, so this doesn't appear necessary. I'll remove it for consistency

scottopell committed:
Removes un-necessary newline on log statement

scottopell committed:
Switches Selection and Ordering enums to iota ints instead of strings

scottopell committed:
Stat each file only once, microbenchmarks show somewhere between 1.3x and 3x improvement depending on number of files

scottopell committed:
Simpler wildcard behavior in fileprovider public interface

scottopell committed:
Adds test with invalid file specified for applyOrdering

scottopell committed:
Refactor FileProvider#FilesToTail to be more explicit, some other small renames

scottopell committed:
Renames config options and adds more description of what each option does

scottopell committed:
Adds release notes for the new config option

scottopell committed:
Removes many-files test that will be flakey across platforms (depends on open file limits)

scottopell commented:
I've addressed all feedback here, thanks to all for the reviews.

@djmitche I missed your question, but all the flags here are permanent, I don't see us ever making `by_modification_time` the default due to the performance differences.

vickenty commented on pkg/config/config.go:
```diff
@@ -893,6 +893,25 @@ func InitConfig(config Config) {
 	// Time in seconds
 	config.BindEnvAndSetDefault("logs_config.file_scan_period", 10.0)
 
+	// Controls how wildcard file log source are prioritized when there are more files
```
Since we're adding a release note for this, would it make sense to add this to `config_template.yaml`? Maybe move the docs there as well, so that they are visible to users.

vickenty commented on pkg/logs/internal/launchers/file/launcher.go:
```diff
@@ -122,10 +135,15 @@ func (s *Launcher) cleanup() {
-	files := s.fileProvider.filesToTail(s.activeSources)
+	files := s.fileProvider.FilesToTail(s.activeSources)
 	filesTailed := make(map[string]bool)
 	tailersLen := len(s.tailers)
 
+	log.Debugf("Scan - got %d files from FilesToTail and currently tailing %d files\n", len(files), tailersLen)
```
tailersLen is used just once before it is re-assigned. Would it make sense to declare variable just then?

```suggestion
	log.Debugf("Scan - got %d files from FilesToTail and currently tailing %d files\n", len(files), len(s.tailers))
```

vickenty commented on pkg/logs/internal/launchers/file/launcher.go:
```diff
@@ -182,6 +191,27 @@ func (s *Launcher) scan() {
 	}
+
+	// Recompute 'tailersLen' to take into account the tailers we stopped.
+	tailersLen = len(s.tailers)
```
```suggestion
	tailersLen := len(s.tailers)
```

vickenty commented on pkg/logs/internal/launchers/file/launcher.go:
```diff
@@ -203,7 +233,11 @@ func (s *Launcher) removeSource(source *sources.LogSource) {
 // launch launches new tailers for a new source.
 func (s *Launcher) launchTailers(source *sources.LogSource) {
-	files, err := s.fileProvider.collectFiles(source)
+	// If we're at the limit already, no need to do a 'CollectFiles', just wait for the next 'scan'
```
What if the new file has a more recent mtime than the ones we're tailing already?

scottopell replied:
I think the only solution in that case would be to manually trigger a `scan` to consider all log sources.

If some file matching the new source has a newer mtime than the currently monitored set, we still need to re-evaluate _all_ matching file mtimes in order to determine what the top N are, what if some file matching a different source has an even newer mtime?

I decided against triggering a scan here and opted to let the next scheduled `scan` take care of this.

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,324 @@
+			} else {
+				files, err := p.CollectFiles(source)
+				if err != nil {
+					if isWildcardSource {
```
`isWildcardSource` is always false here.

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,324 @@
+	// slots as is possible from that source before proceeding to the next one
+	greedySelection selectionStrategy = iota
+	// globalSelection will consider files from all sources together and will choose the
+	// top `filesLimit` files based on the `sortMode` ordering
```
`sortMode` is only mentioned in the comments. Does it need to be updated?

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,324 @@
+// NewFileProvider returns a new Provider
+func NewFileProvider(filesLimit int, wildcardSelection WildcardSelectionStrategy) *FileProvider {
+	wildcardOrder := wildcardReverseLexicographical
+	selectionMode := greedySelection
```
A very minor nit, but would it make sense to remove this at all and use `wildcardSelection` instead? This would allow `applyOrdering` to be split into two functions (it is already called from inside mode-specific branches), allow better name and will make tests simpler (it wouldn't need a fileprovider instance).

scottopell replied:
Happy to split `applyOrdering` into 2 functions, which would indeed make the tests simpler by removing the need for a fileprovider.

To me, separating the state into `wildcardOrder` and `selectionMode` is useful for readability, specifically the `global` vs `greedy` distinction in `FilesToTail`.

vickenty commented on pkg/logs/internal/launchers/file/provider/file_provider.go:
```diff
@@ -0,0 +1,324 @@
+			return statResults[files[i]].After(statResults[files[j]])
+		})
+	} else if p.wildcardOrder == wildcardReverseLexicographical {
+		// FIXME - this codepath assumes that the 'paths' will arrive in lexicographical order
```
Should this `FIXME` be addressed before merging?

scottopell committed:
Adds file_wildcard_selection_mode to config_template and moves docs there

scottopell committed:
Declare tailersLen when its used, not before

scottopell committed:
Updates stale references to 'sort_mode'

scottopell committed:
Removes dead branch of code

scottopell committed:
Splits applyOrdering* functions for simpler logic and testing

kayayarai commented on releasenotes/notes/file-tailing-wildcard-match-mode-73f58cbf7b2a0f73.yaml:
```diff
@@ -0,0 +1,17 @@
+---
+enhancements:
+  - |
+    A new config option ``logs_config.file_wildcard_selection_mode`` has been added.
+    This option allows users to configure how log wildcard file matches are
+    prioritized if the number of matches exceeds ``logs_config.open_files_limit``
+
+    This defaults to ``by_name`` which is the current behavior.
+    The new option is ``by_modification_time`` which will prioritize more recently
+    modified files, but it comes with a performance penalty compared to ``by_name``.
```
```suggestion
    A new config option, ``logs_config.file_wildcard_selection_mode``, 
    allows you to configure how log wildcard file matches are
    prioritized if the number of matches exceeds ``logs_config.open_files_limit``.

    The option defaults to ``by_name`` which is the previous behavior.
    The new option is ``by_modification_time`` which prioritizes more recently
    modified files, but using it can result in slower performance compared to using ``by_name``.
```

kayayarai commented on pkg/config/config_template.yaml:
```diff
@@ -1012,6 +1012,27 @@ api_key:
+  # WARNING: 'by_modification_time' is less performant than 'by_name' and will trigger
+  # more disk I/O at the configured wildcard log paths
+  #
+  # file_wildcard_selection_mode: 'by_name'
```
Throughout this block, change ' to ` around keywords

kayayarai reviewed (changes requested):
some formatting and wording-related suggestions

scottopell committed:
Update releasenotes/notes/file-tailing-wildcard-match-mode-73f58cbf7b2a0f73.yaml

Co-authored-by: Kari Halsted <12926135+kayayarai@users.noreply.github.com>

scottopell committed:
config documentation formatting updates

gh123man reviewed (approved):
LGTM 👍 :shipit:

scottopell commented:
Completed some manual testing and everything looks good. Just need another look at the documentation changes -- @kayayarai

kayayarai commented on pkg/config/config_template.yaml:
```diff
@@ -1012,6 +1012,27 @@ api_key:
+  # chosen for tailing.
+  #
+  # WARNING: `by_modification_time` is less performant than `by_name` and will trigger
+  # more disk I/O at the configured wildcard log paths
```
```suggestion
  # more disk I/O at the configured wildcard log paths.
```

kayayarai reviewed (approved):
Added a period. Plus, Lines 1018 through 1032 should start with two `##` not just one. Otherwise lgtm

scottopell committed:
Corrects some formatting in the config template
